### ISSUE TITLE
X-Script-Name header and Python 3.6 error

### ISSUE THREAD

**Issue by lachlanhurst** [mapproxy/mapproxy#336]
*stacktrace*
```
 Traceback (most recent call last):
   File "/usr/lib/python3.6/threading.py", line 916, in _bootstrap_inner
     self.run()
   File "/usr/lib/python3.6/threading.py", line 864, in run
     self._target(*self._args, **self._kwargs)
   File "/usr/lib/python3.6/socketserver.py", line 641, in process_request_thread
     self.handle_error(request, client_address)
   File "/usr/lib/python3.6/socketserver.py", line 639, in process_request_thread
     self.finish_request(request, client_address)
   File "/usr/lib/python3.6/socketserver.py", line 361, in finish_request
     self.RequestHandlerClass(request, client_address, self)
   File "/usr/lib/python3.6/socketserver.py", line 696, in __init__
     self.handle()
   File "/usr/local/lib/python3.6/dist-packages/mapproxy/util/ext/serving.py", line 217, in handle
     rv = BaseHTTPRequestHandler.handle(self)
   File "/usr/lib/python3.6/http/server.py", line 418, in handle
     self.handle_one_request()
   File "/usr/local/lib/python3.6/dist-packages/mapproxy/util/ext/serving.py", line 252, in handle_one_request
     return self.run_wsgi()
   File "/usr/local/lib/python3.6/dist-packages/mapproxy/util/ext/serving.py", line 194, in run_wsgi
     execute(self.server.app)
   File "/usr/local/lib/python3.6/dist-packages/mapproxy/util/ext/serving.py", line 182, in execute
     application_iter = app(environ, start_response)
   File "/usr/local/lib/python3.6/dist-packages/mapproxy/wsgiapp.py", line 175, in __call__
     req = Request(environ)
   File "/usr/local/lib/python3.6/dist-packages/mapproxy/request/base.py", line 206, in __init__
     if path_info.startswith(script_name):
 TypeError: startswith first arg must be bytes or a tuple of bytes, not str
```

In this case our own proxy service (Tornado based) is passing on a request and including the `X-Script-Name` and `X-Forwarded-Host` headers (see below), but I'd expect this to be reproducible in any MapProxy + Python 3.6 environment. The request being made is a `GetCapabilities`.

```
headers = tornado.httputil.HTTPHeaders()
headers.add('X-Script-Name', '/mapproxy')
headers.add('X-Forwarded-Host', '127.0.0.1')
```

## Quick fix
Replace line [206 of `mapproxy/request/base.py`](https://github.com/mapproxy/mapproxy/blob/master/mapproxy/request/base.py#L206)
```
    if path_info.startswith(script_name):
```
with
```
   if path_info.startswith(script_name.encode()):
```
A log statement shows that `path_info` is a `<class 'bytes'>`. A quick search shows this error message comes up a bit in Python3 migrations. A little unsure if this is an error I've somehow caused on my end.

**Note:** I expect this fix/hack would *break* Python 2 compatibility.

## Additional details
Ubuntu 17.10 (within Docker container)
Python 3.6.3
MapProxy 1.10.4

*requirements.txt*
```
beautifulsoup4==4.6.0
chardet==3.0.4
cryptography==1.9
html5lib==0.999999999
idna==2.5
keyring==10.4.0
keyrings.alt==2.2
lxml==4.0.0
MapProxy==1.10.4
numpy==1.12.1
olefile==0.44
Pillow==4.1.1
pycrypto==2.6.1
pygobject==3.24.1
pyxdg==0.25
PyYAML==3.12
SecretStorage==2.3.1
Shapely==1.6.0
six==1.10.0
webencodings==0.5
```

Thanks!

**Comment by olt** [comment]
I think this is an issue with Tornado. The path_info variable, coming from environ['PATH_INFO'] should be a str(ing) and not bytes.

**Comment by lachlanhurst** [comment]
Possibly, although I was able to reproduce the same issue using `wget`.
```
wget -d --header "X-Script-Name: /mapproxy" \
        --header "X-Forwarded-Host: 127.0.0.1" \
        http://mapproxy:8080/service?REQUEST=GetCapabilities
```

I'll have more of a look into this later on. FWIW I've attached the Dockerfile used here.
[Dockerfile.txt](https://github.com/mapproxy/mapproxy/files/1459568/Dockerfile.txt)

**Comment by lachlanhurst** [comment]
Think I traced down the problem to the following [function](https://github.com/mapproxy/mapproxy/blob/master/mapproxy/util/ext/serving.py#L35). The `str.encode()` function returns a different type in Python2 vs 3.
```
def wsgi_encoding_dance(s, charset='utf-8', errors='replace'):
    if isinstance(s, bytes):
        return s
    return s.encode(charset, errors)
```

In Python2
```
> print(type("test".encode('utf-8', 'replace')))
<type 'str'>
```
In Python3
```
> print(type("test".encode('utf-8', 'replace')))
<class 'bytes'>
```

`wsgi_encoding_dance` is passed a string when [setting up the `environ` dict](https://github.com/mapproxy/mapproxy/blob/master/mapproxy/util/ext/serving.py#L85), and in Python 3 this string will be encoded into a `bytes` object.

**Comment by olt** [comment]
Thanks for tracing it down. ``wsgi_encoding_dance`` is only called when using ``mapproxy-util serve-develop``. So it looks like MapProxy has a similar bug as Tornado.

``wsgi_encoding_dance`` should return a ``str`` according to PEP3333 and I will prepare a fix for that.

Your script might work if you pass the header as a byte, e.g. ``headers.add('X-Script-Name', b'/mapproxy')``. But you should try to fix this in Tornado.

**Comment by olt** [comment]
``wsgi_encoding_dance`` is fixed with ed9851f and your ``wget`` example should work.

**Comment by lachlanhurst** [comment]
Awesome, thanks @olt .

I checked out the latest commit and everything works as expected. Happy for this issue to be closed.
